### PR TITLE
Add measurements

### PR DESCRIPTION
**Pull request by martong**

### CODE REVIEW

**Comment by martong** [comment]
Laptop class cpu can be locked to a specific frequency: https://github.com/martong/finstrument_mock/blob/master/measure/cpu_change_freq.sh

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "googletest"]
 	path = googletest
 	url = https://github.com/google/googletest.git
+[submodule "tbb"]
```
Comment (iszekeres-so) [inline .gitmodules]: Really nice!

File: measurements/measure.cpp [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+#include <thread>
+#include <vector>
+
+#include <tbb/queuing_rw_mutex.h>
+#include <tbb/spin_rw_mutex.h>
+
+#include <urcu-pointer.h>
+#include <urcu.h>
+
+class XRcuPtr {
```
Comment (iszekeres-so) [inline measurements/measure.cpp]: Why the `X` prefix?

File: measurements/measure.cpp [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    void update_all(int value) {
+        tbb::spin_rw_mutex::scoped_lock lock{m}; // write lock
+        for (auto& e : v) {
+            e = value;
+        }
+    }
+};
+
+class XURCU {
+    std::vector<int>* v;
```
Comment (iszekeres-so) [inline measurements/measure.cpp]: How about using `std::unique_ptr<>` instead?

File: measurements/measure.cpp [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        rcu_read_unlock();
+        return result;
+    }
+
+    void update_all(int value) {
+        std::lock_guard<std::mutex> lock{m}; // support concurrent writers
+
+        rcu_read_lock();
+        std::vector<int>* local_copy = rcu_dereference(v);
+        std::vector<int>* local_deep_copy = new std::vector<int>(*local_copy);
```
Comment (iszekeres-so) [inline measurements/measure.cpp]: Same here.
Would it not simplify the code immensely?

File: measurements/measure.cpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    }
+
+    ~XURCU() {
+        synchronize_rcu();
+        delete v;
+    }
+};
+
+class RoundRobin {
+    long long i = 0;
```
Comment (iszekeres-so) [inline measurements/measure.cpp]: How about `std::size_t` for both?

**Comment by iszekeres-so** [comment]
I love it.
How about after merging I'll start working on `google benchmark`s in addition?
Would you like it?

File: measurements/measure.cpp (hunk 2 above)
Comment (martong) [reply measurements/measure.cpp]: X was the name of the original class in the running example. Perhaps we could have `Client` as prefix, or no prefix at all?

File: measurements/measure.cpp (hunk 3 above)
Comment (martong) [reply measurements/measure.cpp]: I am not sure if we could use it, especially `rcu_assign_pointer` seems problematic with the smart ptr

File: measurements/measure.cpp (hunk 5 above)
Comment (martong) [reply measurements/measure.cpp]: Ok

**Comment by martong** [comment]
Yeah, Google benchmark could be useful, thanks for the review!

**Pull request merged**